CONTRIBUTING.md

```diff
@@ -333,6 +333,137 @@ There are several steps that go into a major release
 
 * Update chain JSON schema's recommended versions in `chain.schema.json` located in the root directory.
 
+## Patch and Minor Releases
+
+### Backport Flow
+
+For patch and minor releases, we should already have
+a release branch available in the repository. For example,
+for any v11 release, we have a `v11.x` branch.
+
+Therefore, for any change made to the `main` and, as long as it is
+**state-compatible**, we must backport it to the last major release branch e.g.
+`v11.x` when the next major release is v12.
+
+This helps to minimize the diff of a major upgrade review process.
+
+Additionally, it helps to safely and incrementally test
+state-compatible changes by doing smaller patch releases. Contrary
+to a major release, there is always an opportunity to safely downgrade
+for any minor or patch release.
+
+### State-compatibility
+
+It is critical for the patch and minor releases to be state-machine compatible with
+prior releases in the same major version. For example, v11.2.1 must be
+compatible with v11.1.0 and v11.0.0.
+
+This is to ensure **determinism** i.e. that given the same input, the nodes
+will always produce the same output.
+
+State-incompatibility is allowed for major upgrades because all nodes in the network
+perform it at the same time. Therefore, after the upgrade, the nodes continue
+functioning in a deterministic way.
+
+The state compatibility is ensured by taking the app hash of the state and comparing it
+to the app hash with the rest of the network. Essentially, an app hash is a hash of
+hashes of every store's Merkle root.
+
+For example, at height n, we compute:
+`app hash = hash(hash(root of x/epochs), hash(root of  x/gamm)...)`
```

Comment:
We should also probably note the `LastResultsHash` for tx results matters as well: https://github.com/tendermint/tendermint/blob/main/proto/tendermint/types/types.proto#L70-L77

(The other hashes are unlikely to be an issue / you don't really have to separately think about)

The LastResultsHash today contains:
https://github.com/tendermint/tendermint/blob/main/types/results.go#L47-L54
* Tx GasWanted
* Tx GasUsed
* Tx response "Data"
* Tx response "Code"

The data field includes the proto message response (therefore we cannot change these in patch releases). Tbh idk what the Code field does, as far as I can tell its not even used: https://github.com/cosmos/cosmos-sdk/blob/main/baseapp/abci.go#L298-L304

The gas used field being merkelized means that:
* right now we cannot change the order of operations within a tx execution (different gas used in edge cases of out of gas'ing)
  * Therefore we must be worried about refactoring errors earlier in the tx stack
* We cannot change gas usage amounts (e.g. removing extraneous store reads)

Gas used being merkelized is slated for removal in a subsequent Tendermint release, at which point we will have more flexibility in reordering operations / erroring.

Reply:
Thanks! Great points

re: code

Apparently, it seems to be used when an error is returned. I discussed this briefly here:
https://github.com/osmosis-labs/osmosis/pull/2514#discussion_r956421045

Planning to include a few sentences on it.